Add headless runner functionality with service and implementation

abdelaziz-mahdy opened the pull request:
- Introduced HeadlessRunner and HeadlessRunnerService for managing headless flow execution.
- Implemented HeadlessRunnerServiceImpl to handle runner lifecycle and process management.
- Updated Console.js to support headless mode and integrate with the new service.
- Modified pom.js and services.jrl to include headless runner service in the project structure.

abdelaziz-mahdy committed:
Add headless runner functionality with service and implementation

- Introduced HeadlessRunner and HeadlessRunnerService for managing headless flow execution.
- Implemented HeadlessRunnerServiceImpl to handle runner lifecycle and process management.
- Updated Console.js to support headless mode and integrate with the new service.
- Modified pom.js and services.jrl to include headless runner service in the project structure.

jlhughes commented on src/foam/core/reflow/headless/HeadlessRunnerServiceImpl.java:
```diff
@@ -0,0 +1,97 @@
+import java.util.Map;
+import java.util.UUID;
+
+public class HeadlessRunnerServiceImpl extends ContextAwareSupport implements HeadlessRunnerService {
```
An alternate naming convention is 
HeadlessRunnerServiceServer as we often explicitly have both Server and Client implementation of foam interfaces. 
or simple a HeadlessRunner interface 
and then HeadlessRunnerServer and the cspec is 'headlessRunner' which works for both server and client side. 
'Impl' is a very Java naming convention.

abdelaziz-mahdy replied:
so `HeadlessRunnerService`, should be good right?

then i dont have server and client.

jlhughes replied:
Yes.

jlhughes replied:
ah, also realized the separation you have: foam interfaces can have properties and default method implementations as well. So you can combine the two interfaces into HeadlessRunner, and then HeadlessRunnerService can have any additional logic and method logic which doesn't make sense in the default interface implementations.

abdelaziz-mahdy replied:
oh, I didnt know that

abdelaziz-mahdy replied:
currently HeadlessRunner is a class holder for instance data, and looking at all the services in foam ,none of them define methods in them, they all define it in java file 

so I will just update the name, i hope thats okay.

jlhughes replied:
That's because all the services are older - pre interface supporting properties and methods. 
re: just name change - yep. fine.  Thanks for following up.

abdelaziz-mahdy committed:
Add factory method to schedule property for automatic CronSchedule creation

abdelaziz-mahdy committed:
Update copyright information in HeadlessRunnerService files

abdelaziz-mahdy committed:
Update copyright year to 2025 in HeadlessRunnerService files

abdelaziz-mahdy committed:
Refactor headless runner service: rename implementation class and update service script

abdelaziz-mahdy committed:
Merge branch 'development' of https://github.com/kgrgreer/foam3 into auto-close

kgrgreer commented on src/foam/core/reflow/Flow.js:
```diff
@@ -197,7 +197,10 @@ foam.CLASS({
       of: 'foam.core.cron.CronSchedule',
-      documentation: 'Schedule to run this flow.'
+      documentation: 'Schedule to run this flow.',
+      factory: function() {
```
Is this necessary? Doesn't it do this already?

abdelaziz-mahdy replied:
for some reason it was breaking for me without it, so will test again to be safe

Nauna replied:
If this is going to set a schedule on all flows that means all flows will run on a scheduled cron. Very much undesired.

abdelaziz-mahdy replied:
I was getting an error with dates but it may be because of something else, should we have a flag is it scheduled or not, that will fix up the two use cases and be more clear to the user which is working and which is disabled.

From the user point of view they all have schedules.

abdelaziz-mahdy replied:
done.

kgrgreer commented on src/foam/core/reflow/headless/HeadlessRunnerService.js:
```diff
@@ -0,0 +1,92 @@
+      args: 'String flowName, String sessionId'
+    },
+    {
+      name: 'removeRunner',
```
Did you try calling browser.close().

abdelaziz-mahdy replied:
yes and it didnt work.

abdelaziz-mahdy replied:
for context 

 `ReferenceError: browser is not defined` and checking the resources says its available from chrome apps

Nauna committed:
Refactor headless mode logging and update schedule property to use IntervalSchedule

Nauna reviewed (commented):
How is (headlessRunnerService) different then using a threadPool?

abdelaziz-mahdy commented:
> How is (headlessRunnerService) different then using a threadPool?
> 

I don't think we have client side access for thread pool, and exposing it didn't seem like a good idea from my side, so I went with a separation of concern approach.

Nauna commented on src/foam/core/reflow/Flow.js:
```diff
@@ -196,8 +196,13 @@ foam.CLASS({
       class: 'FObjectProperty',
-      of: 'foam.core.cron.CronSchedule',
-      documentation: 'Schedule to run this flow.'
+      of: 'foam.core.cron.IntervalSchedule',
```
This change is removing the script from our test flow... so just documenting this here:

> [
	{
"flowName": "script1",
"cmd": "script",
"value": {
"class": "foam.core.reflow.Script",
"code": "n = foam.core.notification.Notification.create({ broadcasted:true, body:'hello', toastMessage:'toast' }); this.__context__.notificationDAO.put(n).then(function() { log('done'); });"
		}
	},
	{
"flowName": "a1",
"cmd": "script1.run()"
	}
]

abdelaziz-mahdy replied:
Noted.

kgrgreer replied:
I don't think we want IntervalSchedule. It doesn't match our use case.
We could allow them to pick different scheduling strategies with:
      of: 'foam.core.cron.Schedule',
      view: {
        class: 'foam.u2.view.FObjectView',
        of: 'foam.core.cron.Schedule'
      },

abdelaziz-mahdy replied:
done.

abdelaziz-mahdy committed:
Change schedule property to use Schedule class.
This allows for more flexible scheduling options
and improves the overall flow management.

jlhughes commented:
What is the idea here, simply to 'schedule' execution of a flow? If so I suggest a different approach.
Consider a  flowDAO decorator, FlowJobDAO, which on put (create/update), checks if a flow has a schedule, if so create a FlowJob (another model which extends 'foam.core.cron.Schedulable') and save to schedulableDAO.  On update, if schedule removed, remove the flowjob.  Also on update, if schedule was one time and it is past the scheduled time, then again clean up and remove the flowjob. 
This additional model, FlowJob, is so we don't have to make Flow extends Schedule/Cron. 
This model will have properties for the flow id of interest, plus, userid, sessionid. 
The FlowJobDAO can have properties such as userId and sessionId that are used to configure each FlowJob.  You'll also create a user and session (with a long ttl), that only the flowjobs will use. 
The FlowJob will also implement 'execute', and its here that you execute your flow.  
@noodlemoodle wrote Schedulable, so we can both help with this. 
Summary:
FlowJob, extends Schedulable, implements execute (of Cron/Script). 
FlowJobDAO, decorates flowDAO, and manages the flowjobs. 
The only change to flow is to add the 'schedule' property.  (this schedule will get copied to the FlowJob). 
If @kgrgreer doesn't want 'schedule' on Flow, then you could create an 'action' which could popup and take a schedule as input, then create the flowjob and put to flowJobDAO which has a FlowJobDAO decoration which finalizes the flowjob setup.

abdelaziz-mahdy commented:
> What is the idea here, simply to 'schedule' execution of a flow? If so I suggest a different approach. Consider a flowDAO decorator, FlowJobDAO, which on put (create/update), checks if a flow has a schedule, if so create a FlowJob (another model which extends 'foam.core.cron.Schedulable') and save to schedulableDAO. On update, if schedule removed, remove the flowjob. Also on update, if schedule was one time and it is past the scheduled time, then again clean up and remove the flowjob. This additional model, FlowJob, is so we don't have to make Flow extends Schedule/Cron. This model will have properties for the flow id of interest, plus, userid, sessionid. The FlowJobDAO can have properties such as userId and sessionId that are used to configure each FlowJob. You'll also create a user and session (with a long ttl), that only the flowjobs will use. The FlowJob will also implement 'execute', and its here that you execute your flow. @noodlemoodle wrote Schedulable, so we can both help with this. Summary: FlowJob, extends Schedulable, implements execute (of Cron/Script). FlowJobDAO, decorates flowDAO, and manages the flowjobs. The only change to flow is to add the 'schedule' property. (this schedule will get copied to the FlowJob). If @kgrgreer doesn't want 'schedule' on Flow, then you could create an 'action' which could popup and take a schedule as input, then create the flowjob and put to flowJobDAO which has a FlowJobDAO decoration which finalizes the flowjob setup.

Currently we are running the flows in headless chrome so we made one cron that runs every 5 mins and checks the schedule and based on it runs the flow (hence we plan to run one flow at a time) and the headless runner objective is to know if and when the flow finishes running so we can terminate the chrome process.

abdelaziz-mahdy committed:
feat: add waitForRunnerRemoval method to handle runner removal with timeout

Nauna reviewed (approved):
LGTM

kgrgreer commented on src/foam/core/reflow/Flow.js:
```diff
@@ -28,7 +28,7 @@ foam.CLASS({
 
   requires: [
-    'foam.core.cron.CronSchedule',
+    'foam.core.cron.IntervalSchedule',
```
Why require IntervalSchedule?

abdelaziz-mahdy committed:
merge with dev

abdelaziz-mahdy committed:
fix: remove unused 'lib' entry from POM configuration
